### PR TITLE
[wip] common: disable creation of noarch packages.

### PR DESCRIPTION
**Pull request by fosslinux**
This change disables the creation of noarch packages.

I believe that most package maintainers are of the same opinion that noarch causes far more problems than it solves. For some time on IRC there has been continued discussion about disabling it. This is what this PR attempts to achieve.

This still supports `archs=noarch` but IMO this is misleading if it dosen't actually create a package with the noarch arch. This is why I have made `archs=noarch` just mean `noarch=yes`. This is how I think all future packages should declare noarch: `noarch=yes`. I also considered `data_only=yes` or `arch_independent=yes` but I don't really like either of those.

The reason noarch even still exists is because it is a useful shorthand for `nodebug=yes`, `nostrip=yes`, etc etc.

The removed 32-bit block about noarch is useless because noarch already forces `noshlibprovides`.

NB: Right now, this is a MVP and still needs some more work.

Pinging @q66 @the-maldridge

### CODE REVIEW

**Comment by ericonr** [comment]
I like noarch because it enables me to build something like the void-docs package and put it in the release so that everyone can download it as a preview. @zdykstra does the same thing with zfsbootmenu, where it's even more useful because he does release candidates.

I guess this use case would be broken here, right?

**Comment by ahesford** [comment]
That `noarch` implies `nodebug`, `nostrip`, `noshlibprovides` and whatever else does not make it a useful shorthand for those things, because it also implies other things that are being disregarded by dropping current `noarch` functionality.

If you remove the current `noarch` support, `noarch` should be a forbidden keyword in `archs` and there should be no `noarch=yes`.

**Comment by ahesford** [comment]
Thinking about the distribution use case mentioned by @ericonr, what if we don't remove support for `noarch`, but move it to a `noarch=yes` flag (which would be mutually exclusive with `archs`) and make `xbps-src` require a configuration option to actually produce a `noarch` package?

This would behave sort of like `restricted` does now, except rather than refusing to build a `noarch=yes` package, `xbps-src` would treat the package as if `noarch` weren't set unless `XBPS_ALLOW_NOARCH=yes` in the environment. Offline distributors could set that option to produce packages that work for everybody, but the official builders can omit the option to produce per-arch packages everywhere.

**Comment by Chocimier** [comment]
> noarch causes far more problems than it solves


@void-linux/pkg-committers, let's start discussion by enumerating them. Most obvious ones, please complement:

Advantages of noarch:
+ Less copies of big data packages on mirrors
+ Ability to create packages installable anywhere, used for testing xbps or experimental pure data packages as described by @ericonr 
+ Marking for all related hooks at once (nodebug, ...)

Disadvantages of noarch:
- More complicated cleaning old versions of packages shared between repodatas on primary mirror

**Comment by q66** [comment]
I strongly disagree with changing anything in templates themselves, there is just no need for that

the only change we need is making xbps-create always make arched packages, templates can stay exactly the same

**Comment by ahesford** [comment]
I don't mind preserving the `archs=noarch` construct in templates and disabling it, but I still think it would be nice to toggle this with an `XBPS_ALLOW_NOARCH` env var to allow people to make `noarch` packages locally for distribution.

**Comment by q66** [comment]
i don't think we need to or that we should in the first place, the use case described by @ericonr is unsupported

**Comment by ahesford** [comment]
It's one thing to make things needless complicated to support some unsupported, nonstandard use. It's another thing to make simple accommodations that preserve existing workflows. Unless I'm missing something, the accommodation looks as simple as changing
```
if [ "${archs// /}" = "noarch" ]; then
```
to
```
if [ "${archs// /}" = "noarch" -a -n "$XBPS_ALLOW_NOARCH" ]; then
```
at https://github.com/void-linux/void-packages/blob/master/common/hooks/do-pkg/00-gen-pkg.sh#L107

**Comment by q66** [comment]
and then we drop noarch support from xbps and what? we shouldn't be encouraging non-standard uses that we don't wish to support

**Comment by ericonr** [comment]
> and then we drop noarch support from xbps and what? 

Would there be a time frame on this? At some point we'd have to rebuild all the missing noarch packages, like `0ad-data`.

**Comment by q66** [comment]
there is no timeframe, but considering how broken the noarch specific code in xbps is, it's bound to happen at one point

obviously existing noarch packages would have to be rebuilt, but that's no problem, it could just be done ahead of time

**Comment by fosslinux** [comment]
Ok, replying to some comments here:

@ericonr: 
> I like noarch because it enables me to build something like the void-docs package and put it in the release so that everyone can download it as a preview. @zdykstra does the same thing with zfsbootmenu, where it's even more useful because he does release candidates.

What do you mean? I don't really understand this use-case. Is the idea you build once and then `XBPS_TARGET_ARCH=x86_^4-musl xbps-rindex -a *.xbps` for each architecture? Doing things like this is completely unsupported AFAICT.

@ahesford 
> That noarch implies nodebug, nostrip, noshlibprovides and whatever else does not make it a useful shorthand for those things, because it also implies other things that are being disregarded by dropping current noarch functionality.

I don't quite get what you mean. These other things are unnessecary, and the shorthand for `nodebug`, etc is useful for data packages. This is why I think we shouldr ename it.

**Comment by the-maldridge** [comment]
Primary advantages with noarch
  * it saves a meager amount of disk space that we're somehow still worried about in 2020 when you can buy commodity 10T disks at reasonable prices delivered to your DC of choice in around 2 days.

Primary disadvantages of noarch
  * pretty much everything else about it.

In no particular order noarch relies on kludges within xbps itself, it makes managing the repos unnecessarily complicated, it makes it shockingly difficult to construct a coherent package graph that could be used to schedule builds in a distributed fashion (along with our kludgy multilib, but that's another story).

As far as timeline to drop noarch from xbps?  As soon as it is practically feasible.  Removing that code would be on the critical path to improving the dependency solver and generally making xbps more robust.  I would go so far as to propose a bulk rebuild of all current noarch packages to make it possible to dump the support from xbps sooner.

I disagree that there is no need to touch templates.  It makes sense to remove the components that trigger this behavior to minimize confusion.  We are not intending to support it, so we should not keep the detritus in place that suggests we do.

**Comment by fosslinux** [comment]
Agree with everything above.

The touching templates part I agree with to some extent, I don't really care if we rebuild noarch packages, or if we do a bulk change on noarch templates, etc.

The two main questions here I see that are unresolved:

1. Should we support noarch, in any capacity?
2. Should we have some kind of shorthand for useful variables such as nostrip for data packages?
3. Should we immediatley rebuild all pacakages?

I say 1) no 2) yes 3) don't really mind, but sooner is better so yes

**Comment by the-maldridge** [comment]
A data package shouldn't need the shorthand, because there will be nothing to strip in it.  That sounds like something that isn't actually a data package.

**Comment by q66** [comment]
it's less about having things in it and more about not taking the time scanning for these things (noarch packages can contain many files and needlessly checking everything can be slow)

so i still think being able to describe the intent in templates is fine, it's just that xbps itself shouldn't support it (i.e. for each noarch template, an arched package is generated)

**Comment by fosslinux** [comment]
> not taking the time scanning for these things

I feel like the actual solution to this is optimising strip/debug hooks in xbps-uhelper

**Comment by sgn** [comment]
> > not taking the time scanning for these things
> 
> I feel like the actual solution to this is optimising strip/debug hooks in xbps-uhelper

I still take time to read all files to memory, and the check is in `common/hooks/post-install/06-strip-and-debug-pkgs.sh`

Yes, that hook could be improved.

**Comment by ericonr** [comment]
> (noarch packages can contain many files and needlessly checking everything can be slow)

@q66 I think being more strict re. checks would be a good idea, and speeding up certain checks is definitely doable once we have `xbps-uhelper` doing more things. Since I don't know the ideal shape for that, it's hard to get started on any part of it, but I would like to help there.

> What do you mean? I don't really understand this use-case. Is the idea you build once

@fosslinux see the packages here https://github.com/void-linux/void-docs/releases

**Comment by sgn** [comment]
Anyway, `file(1)` is able to handle batch of files in `common/hooks/post-install/06-strip-and-debug-pkgs.sh`,
no need to polute `xbps-uhelper(1)`.
Or, did @fosslinux talk about something else?

**Comment by fosslinux** [comment]
@sgn no, there has been some discussion about moving the debug and strip checks out to `xbps-uhelper` to speed up performance. @q66 and @ericonr were the ones who originally thought about it

**Comment by ahesford** [comment]
> I don't quite get what you mean. These other things are unnessecary, and the shorthand for `nodebug`, etc is useful for data packages. This is why I think we should rename it.

That a package is `noarch` implies that it has no native executables (which I am distinguishing from text files that are parsed by an interpreter and have an execute bit set). That a package has no native executables implies that we can set `nodebug` and `nostrip`. Transitively, `noarch` implies these things too, but if we're going to lop out the core functionality invoked by `noarch`, we shouldn't keep some vestigial keyword around just for convenience.

Keeping a `noarch` keyword without supporting `noarch` packages, where `noarch` is a wildcard for the package manager, is just misdirection. The right way to signal that we don't need to debug or strip anything is to add a `noexecutable` flag or something similarly named, or else just add the individual flags to templates. Calling it `noarch` is just going to lead to confusion, and keeping `archs=noarch` produces unexpected results because it has specific meaning now.

I haven't thought about package dependency graphs and don't much care to. In principle, it doesn't seem like properly supporting `noarch` in dependency resolution (and it seems to me the repo-maintenance challenges are just dependency resolution across the architecture dimension) should be too challenging, but I can certainly believe that difficulties in the implementation quickly become more trouble than they are worth to save some disk space.

If that's the case, drop `noarch` from XBPS, modify `xbps-src` to treat `archs=noarch` as `noexecutable=yes` and emit a warning about this transitional reinterpretation, modify `xlint` to complain when `arches=noarch` is detected and recommend the new keyword, and once we've rebuilt every `noarch` package with `noexecutable=yes` instead, drop these transitional modifications. Or just bite the bullet, migrate every template in one pass and be done with it.

**Comment by ericonr** [comment]
> That a package has no native executables implies that we can set nodebug and nostrip.

Ideally `nodebug` and `nostrip` should only ever be set for packages that don't work when they are stripped or debugged, and moving to a `noexecutable` flag which scans the generated package for the wrong files would be a nice move. Using flags to speed up build checking isn't a valid use case, once we implement better helpers.

**Comment by the-maldridge** [comment]
There's no need to have a transitional warning for the templates.  -packages is a monorepo which allows us to refactor things in one go rather than hanging on to things for years with linter warnings.  This isn't debian...

**Comment by fosslinux** [comment]
After some discussion on IRC. I believe the best way forward is to drop *any* distinction between data and normal packages, i.e. drop noarch, no `noexecutable=yes`, nothing like that, and instead improve the strip/debug/shlibprovides hooks to be faster. I am working on the strip hook now, there's some pretty clear optimizations that can be done.

Also, as per @the-maldridge 's suggestion above, `archs=noarch` will be removed from every package at once.

I'll fix up this PR very soon.

Do we want to revbump noarch packages, or not?

**Comment by ericonr** [comment]
> Do we want to revbump noarch packages, or not?

I doubt all the packages will just build, so this PR would likely contain a pretty big amount of fixes and updates if we were to revbump/update everything. Not sure what's the best way forward.

**Comment by the-maldridge** [comment]
@ericonr do we want to revbump noarch packages? yes, but not in this patch.

**Comment by fosslinux** [comment]
Yeah, I meant later on, not in this PR.

**Comment by fosslinux** [comment]
@void-linux/pkg-commiters can you guys take a look at this again.

I'm not going to bother continously rebasing the massive srcpkgs commit, I'll just do it once again just before merge.

**Comment by fosslinux** [comment]
Uh, @void-linux/pkg-committers

(editing doesn't reping)

**Review by Piraty — CHANGES_REQUESTED** [review]
Manual.md needs to reflect that as well

File: common/hooks/pre-pkg/06-shlib-provides.sh
Comment (Chocimier) [inline common/hooks/pre-pkg/06-shlib-provides.sh:48]: Equivalent to old condition would be `[ -n "$noshlibprovides" ]`, but I do not get intent of its left side of `-o`.

**Comment by Chocimier** [comment]
Don't forget to revbump packages when dropping noarch in template. If rebuilding at once is too much, do in batches.

**Comment by ericonr** [comment]
@Chocimier revbumping in this PR isn't feasible, because undoubtedly a lot of packages will require fixing / updating. We can make an issue about noarch packages that are still in the repo once this is done, IMO. Removing `noarch` from templates is still necessary, of course.

File: common/hooks/pre-pkg/06-shlib-provides.sh
Comment (Chocimier) [reply common/hooks/pre-pkg/06-shlib-provides.sh:48]: So old left side is "Noarchs packges don't have shared libs, but if $shlib_provides is set, run the hook". Please change to `[ -n "$noshlibprovides" ]`.

Comment (fosslinux) [reply common/hooks/pre-pkg/06-shlib-provides.sh:48]: Oh, yeah, so it is, oops.

**Comment by fosslinux** [comment]
> Don't forget to revbump packages when dropping noarch in template. If rebuilding at once is too much, do in batches.

I plan to do this in a separate PR directly after this PR is merged.

The reason for the drop of `archs=noarch` in all templates is because after the first commit of this PR `archs=noarch` is syntactically incorrect.

I think we can make a distinction between removing noarch from `void-packages`(i.e. what this PR currently does) and removing noarch packages themselves from the mirrors (i.e. revbumping).

**Comment by fosslinux** [comment]
@Piraty Can you check if my Manual.md changes are sound.

File: Manual.md
Comment (Chocimier) [inline Manual.md:917]: Not removed from code.

**Comment by Chocimier** [comment]
I am more concerned of temporary state when same revision is noarch in repo but when rebuild locally not noarch and of losing easy way to see if package was rebuilt than of excessive variable in template.
Rebuilding before dropping support in xbps-src would be cleaner, but if you revbump *directly after* it is fine too.

File: Manual.md
Comment (fosslinux) [reply Manual.md:917]: I swear I did this....

**Comment by fosslinux** [comment]
@Chocimier Fair, I get that point. However, I don't believe this will be the state for very long so it should be fine. I don't see people rebuilding a noarch package that much locally without also updating it...

How is rebuilding before dropping support going to work? Rebuilding them will just generate a new noarch package if we don't drop support first.

**Comment by sgn** [comment]
On 2020-08-27 17:43:55-0700, fosslinux <notifications@github.com> wrote:
> How is rebuilding before dropping support going to work? Rebuilding
> them will just generate a new noarch package if we don't drop
> support first.

Uhm, by removing "archs=noarch" and revbump for building?

**Comment by fosslinux** [comment]
Oh, I understand now, I was unsure about @Chocimier meant.

I am happy to leave this PR for now and do it in that order instead, if there is appropriate consensus.

**Comment by fosslinux** [comment]
In the process of creating a `noarch` revbump commit, will be included in this PR.

Three-stage process:

1. Merge fixing PRs:
- [ ] #24597
2. Merge this PR which will include two commits, the `srcpkgs: remove noarch` PR which will revbump every noarch package followed by the `common` commit which will disable noarch. Unless it is simpler to split it out into two PRs.

File: Manual.md
Comment (fosslinux) [reply Manual.md:917]: (in next push)

**Comment by fosslinux** [comment]
Request: Can we please not accept any new packages with noarch, just not have noarch for any new packages while I prepare the rebuild; it will just make this commit much easier.

Can this be added to the Manual too? Don't really care on the exact wording but something along the lines of:

> We are in the process of transitioning away from noarch. Please no longer use noarch for new packages.

**Comment by ahesford** [comment]
If the transition time will be short, maybe now is the time for an `xlint` rule that complains about `noarch`.

**Comment by fosslinux** [comment]
I agree. Will make a PR for that...

**Comment by fosslinux** [comment]
Am going to remove the noarch rebuild commit from this PR and split it out into various different PRs; doing one massive one is proving to be more trouble than it is worth so I plan to do it in smaller sets of around 100-500.

File: Manual.md
Comment (fosslinux) [reply Manual.md:917]: Actually this will come in the first batch set with revbumps of `build_style=gem` packages.

**Comment by Chocimier** [comment]
> Can this be added to the Manual too? Don't really care on the exact wording but something along the lines of:
> 
> > We are in the process of transitioning away from noarch. Please no longer use noarch for new packages.

Changes to manual from this PR could be merged first as sign of decision taken.

**Comment by fosslinux** [comment]
I have split out changes to the manual into a separate PR, #24684.

Committers, when pushing anything that updates or revbumps something, or merging a PR, please ensure that if it is `archs=noarch` it is removed. This will expedite the noarch removal process.

**Comment by sgn** [comment]
remaining package:

- [ ] binutils #26502 
- [ ] buildbot-slave
- [ ] buildbot
- [ ] cherrytree #26558 
- [ ] geoip-data
- [ ] ghc
- [ ] glibc
- [ ] grub #25267 
- [ ] rust #26609 
- [ ] uhd #26748

**Comment by fosslinux** [comment]
rust @ #27119

**Comment by fosslinux** [comment]
finally

**Pull request closed without merging**
